comb_spec_searcher/equiv_db.py

```diff
@@ -70,16 +71,60 @@ def __getitem__(self, comb_class: int) -> int:
             self.parents[ancestor] = root
         return root
 
-    def add_edge(self, label: int, other_label: int) -> None:
+    def get_one_way_vertices(self) -> Dict[int, Set[int]]:
+        res: Dict[int, Set[int]] = defaultdict(set)
+        for start, ends in self._one_way_vertices.items():
+            start = self[start]
+            for end in ends:
+                end = self[end]
+                if start != end:
+                    res[start].add(end)
+        self._one_way_vertices = res
+        return res
+
+    def add_one_way_edge(self, label: int, other_label: int) -> None:
+        self._add_edge(label, other_label)
+        self._one_way_vertices[self[label]].add(self[other_label])
+
+    @cssmethodtimer("find_paths")
+    def connect_cycles(self):
+        """Look for cycles using one way edges that have been added."""
+        one_way_vertices = self.get_one_way_vertices()
+        dequeue: Deque[Tuple[int, ...]] = deque()
+        dequeue.extend([(label,) for label in one_way_vertices.keys()])
+        visited = set()
+        while dequeue:
+            path = dequeue.popleft()
+            end = path[-1]
+            if end in visited:
+                continue
+            visited.add(end)
+            for new_end in one_way_vertices[end]:
+                for i, vertex in enumerate(path[:-1]):
+                    if self.equivalent(vertex, new_end):
```

Comment:
Since we are already at the level of equivalence, isn't equality check sufficient?

Reply:
We set equivalence as we go, so labels can change.

Reply:
This true